firebase-vscode/webviews/EmulatorPanel.tsx

```diff
@@ -0,0 +1,250 @@
+  firebaseJson,
+}: {
+  firebaseJson: FirebaseConfig;
+}) {
+  if (!firebaseJson) {
+    throw Error("Expected a valid FirebaseConfig.");
+  }
+  const [emulatorUiSelections, setEmulatorUiSelections] =
+    useState<EmulatorUiSelections>(DEFAULT_EMULATOR_UI_SELECTIONS);
+
+  webLogger.debug(
+    "initial state ui selections:" + JSON.stringify(emulatorUiSelections)
+  );
+  function setEmulatorUiSelectionsAndSaveToWorkspace(
+    uiSelections: EmulatorUiSelections
+  ) {
+    // TODO(christhompson): Save UI selections in the current workspace.
+    // Requires context object.
+    setEmulatorUiSelections(uiSelections);
+  }
+  const [showEmulatorProgressIndicator, setShowEmulatorProgressIndicator] =
+    useState<boolean>(false);
+
+  // TODO(christhompson): Load UI selections from the current workspace.
+  // Requires context object.
+  // TODO(christhompson): Check if the emulators are running on extension start.
+  const [runningEmulatorInfo, setRunningEmulatorInfo] =
+    useState<RunningEmulatorInfo>();
+
+  broker.on("notifyEmulatorsStopped", () => {
+    setShowEmulatorProgressIndicator(false);
+    webLogger.debug(`notifyEmulatorsStopped received in sidebar`);
+    setRunningEmulatorInfo(null);
+  });
+
+  broker.on("notifyRunningEmulatorInfo", (info: RunningEmulatorInfo) => {
+    setShowEmulatorProgressIndicator(false);
+    webLogger.debug(`notifyRunningEmulatorInfo received in sidebar`);
+    setRunningEmulatorInfo(info);
+  });
+
+  broker.on("notifyEmulatorImportFolder", ({ folder }) => {
+    webLogger.debug(`notifyEmulatorImportFolder received in sidebar: ${folder}`);
+    emulatorUiSelections.importStateFolderPath = folder;
+    setEmulatorUiSelectionsAndSaveToWorkspace({ ...emulatorUiSelections }); // rerender clone
+  });
+
+  function launchEmulators() {
+    if (!emulatorUiSelections.projectId) {
+      broker.send("showMessage", {
+        msg: "Missing project ID",
+        options: {
+          modal: true,
+          detail: `Please specify a project ID before starting the emulator suite.`,
+        },
+      });
+      return;
+    }
+    if (!firebaseJson) {
+      // TODO(christhompson): Consider using a default config in the case that
+      // firebase.json doesnt exist.
+      broker.send("showMessage", {
+        msg: "Missing firebase.json",
+        options: {
+          modal: true,
+          detail: `Unable to find firebase.json file.`,
+        },
+      });
+      return;
+    }
+    setShowEmulatorProgressIndicator(true);
+    broker.send("launchEmulators", {
+      firebaseJson,
+      emulatorUiSelections,
+    });
+  }
+
+  function stopEmulators() {
+    setShowEmulatorProgressIndicator(true);
+    broker.send("stopEmulators");
+  }
+
+  /**
+   * Called when import folder changes.
+   */
+  function selectedImportFolder(event: any) {
+    event.preventDefault();
+    broker.send("selectEmulatorImportFolder");
+  }
+
+  function toggleExportOnExit() {
+    webLogger.debug("toggle export on exit");
```

Comment:
nit: include what it is being toggled to

Reply:
Done